File: go/v2/crypto/ecdsa.go
// ECDSA-SHA256 RAW 签名/验签（RFC 6979 deterministic）。
//
// AUN V2 §3.1：
//   - 曲线 P-256 / SHA-256
//   - 签名 RAW 编码 r(32B) || s(32B) = 64 字节定长
//   - MUST 使用 RFC 6979 deterministic nonce 以保证字节级可复现
//
// Go 标准库 crypto/ecdsa 不暴露 RFC 6979 入口（SignASN1 输出 ASN.1 DER 且 nonce
// 由 rand.Reader 提供，非确定性）。本文件实现 RFC 6979 §3.2 的 K 生成算法以及
// 配套的 P-256 ECDSA 签名流程，输出与 Python `cryptography` 的
// `deterministic_signing=True` 字节级一致。

package crypto

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/hmac"
	"crypto/sha256"
	"crypto/x509"
	"errors"
	"fmt"
	"hash"
	"math/big"
)

// ECDSARawSigLen RAW 编码签名长度（r 32B || s 32B）
const ECDSARawSigLen = 64

// ECDSASignRaw 使用 RFC 6979 deterministic nonce 在 P-256 上对 message 做
// SHA-256 ECDSA 签名，输出 RAW 编码 (r 32B || s 32B)。
func ECDSASignRaw(privateKeyScalar, message []byte) ([]byte, error) {
	if len(privateKeyScalar) != 32 {
		return nil, fmt.Errorf("ECDSA: 私钥标量长度无效，期望 32 字节，实际 %d", len(privateKeyScalar))
	}
	curve := elliptic.P256()
	n := curve.Params().N

	d := new(big.Int).SetBytes(privateKeyScalar)
	if d.Sign() == 0 || d.Cmp(n) >= 0 {
		return nil, errors.New("ECDSA: 私钥标量超出曲线阶范围")
	}

	hash := sha256.Sum256(message)
	h1 := hash[:]

	// RFC 6979 §3.2: 派生 deterministic k，并完成签名
	r, s := signRFC6979(curve, d, h1)
	if r == nil || s == nil {
		return nil, errors.New("ECDSA: 签名失败（k 迭代未产生有效结果）")
	}

	out := make([]byte, ECDSARawSigLen)
	r.FillBytes(out[:32])
	s.FillBytes(out[32:])
	return out, nil
}

// ECDSAVerifyRaw 验证 RAW 编码的 ECDSA-SHA256 签名。
func ECDSAVerifyRaw(publicKeyDER, signatureRaw, message []byte) bool {
	if len(signatureRaw) != ECDSARawSigLen {
		return false
	}
	pub, err := x509.ParsePKIXPublicKey(publicKeyDER)
	if err != nil {
		return false
	}
	ecPub, ok := pub.(*ecdsa.PublicKey)
	if !ok || ecPub.Curve != elliptic.P256() {
		return false
	}
	r := new(big.Int).SetBytes(signatureRaw[:32])
	s := new(big.Int).SetBytes(signatureRaw[32:])
	hash := sha256.Sum256(message)
	return ecdsa.Verify(ecPub, hash[:], r, s)
}

// signRFC6979 在 P-256 上完成 (k, r, s) 的派生与签名。
// 当 r 或 s 为 0 时重新派生 k（RFC 6979 §3.2 规定）。
func signRFC6979(curve elliptic.Curve, d *big.Int, h1 []byte) (*big.Int, *big.Int) {
	params := curve.Params()
	n := params.N
	// 对 SHA-256 + P-256：qlen = 256，rolen = 32
	const rolen = 32

	// hash function constructor
	newHash := sha256.New
	hashLen := sha256.Size

	// h1 = SHA-256(message)，长度 32B，刚好等于 qlen/8。
	// bits2octets(h1) = int2octets(bits2int(h1) mod q)
	h1Int := bits2int(h1, params.BitSize)
	h1Mod := new(big.Int).Mod(h1Int, n)
	h1Octets := int2octets(h1Mod, rolen)
	xOctets := int2octets(d, rolen)

	// V = 0x01 重复 hashLen 次；K = 0x00 重复 hashLen 次
	V := bytes_repeat(0x01, hashLen)
	K := bytes_repeat(0x00, hashLen)

	// K = HMAC_K(V || 0x00 || x_octets || h1_octets)
	K = hmacSum(newHash, K, V, []byte{0x00}, xOctets, h1Octets)
	V = hmacSum(newHash, K, V)
	K = hmacSum(newHash, K, V, []byte{0x01}, xOctets, h1Octets)
	V = hmacSum(newHash, K, V)

	for {
		// 生成 T，长度 >= qlen
		T := make([]byte, 0, rolen)
		for len(T) < rolen {
			V = hmacSum(newHash, K, V)
			T = append(T, V...)
		}
		k := bits2int(T, params.BitSize)
		if k.Sign() > 0 && k.Cmp(n) < 0 {
			// 计算 r = (k*G).x mod n
			kBytes := int2octets(k, rolen)
			x, _ := curve.ScalarBaseMult(kBytes)
			r := new(big.Int).Mod(x, n)
			if r.Sign() != 0 {
				// s = k^-1 * (h + r*d) mod n
				kInv := new(big.Int).ModInverse(k, n)
				if kInv != nil {
					rd := new(big.Int).Mul(r, d)
					rd.Mod(rd, n)
					sum := new(big.Int).Add(h1Mod, rd)
					sum.Mod(sum, n)
					s := new(big.Int).Mul(kInv, sum)
					s.Mod(s, n)
					if s.Sign() != 0 {
						return r, s
					}
				}
			}
		}
		// 失败：迭代 K, V
		K = hmacSum(newHash, K, V, []byte{0x00})
		V = hmacSum(newHash, K, V)
	}
}

// bits2int 取 in 的左侧 qlen 位（不足则补 0），转为大整数。
// 当 in 长度 * 8 > qlen 时，右移多余位。
func bits2int(in []byte, qlen int) *big.Int {
	v := new(big.Int).SetBytes(in)
	vlen := len(in) * 8
	if vlen > qlen {
		v.Rsh(v, uint(vlen-qlen))
	}
	return v
}

// int2octets 将 v 编码为定长 rolen 字节大端。
func int2octets(v *big.Int, rolen int) []byte {
	out := make([]byte, rolen)
	v.FillBytes(out)
	return out
}

func bytes_repeat(b byte, n int) []byte {
	out := make([]byte, n)
	for i := range out {
		out[i] = b
	}
	return out
}

// hmacSum 计算 HMAC(key, concat(parts...))。
func hmacSum(newHash func() hash.Hash, key []byte, parts ...[]byte) []byte {
	mac := hmac.New(newHash, key)
	for _, p := range parts {
		mac.Write(p)
	}
	return mac.Sum(nil)
}

File: ts/src/v2/crypto/ecdsa.ts
/**
 * AUN E2EE V2: ECDSA-SHA256 RAW（RFC 6979 deterministic）
 *
 * 规范引用：§3.1
 * - 曲线 P-256
 * - RFC 6979 deterministic nonce（必须）
 * - 输出 RAW 编码 r(32B)||s(32B)
 *
 * 实现选型：
 * - 使用 @noble/curves/nist 提供 deterministic ECDSA
 * - 公钥使用 DER SubjectPublicKeyInfo（与 Python/Go 对齐）
 *   通过 Node createPublicKey 提取未压缩点（0x04||X||Y）后传给 noble
 */

import { p256 } from '@noble/curves/nist.js';
import { createHash, createPublicKey } from 'node:crypto';

/**
 * 把 base64url 字符串解码为 32 字节大端序无前导零填充的字节数组。
 */
function b64uToFixed32(b64url: string): Buffer {
  const buf = Buffer.from(b64url, 'base64url');
  if (buf.length === 32) return buf;
  if (buf.length < 32) {
    const padded = Buffer.alloc(32);
    buf.copy(padded, 32 - buf.length);
    return padded;
  }
  const start = buf.length - 32;
  for (let i = 0; i < start; i++) {
    if (buf[i] !== 0) {
      throw new Error(`invalid EC coordinate: length=${buf.length}, leading non-zero byte`);
    }
  }
  return buf.subarray(start, buf.length);
}

/**
 * 从 DER SPKI 编码公钥提取未压缩点（0x04||X||Y），共 65 字节。
 */
function derSpkiToUncompressed(publicKeyDer: Uint8Array): Uint8Array {
  const pub = createPublicKey({
    key: Buffer.from(publicKeyDer),
    format: 'der',
    type: 'spki',
  });
  const jwk = pub.export({ format: 'jwk' });
  if (jwk.kty !== 'EC' || jwk.crv !== 'P-256' || !jwk.x || !jwk.y) {
    throw new Error('public key is not a P-256 EC key');
  }
  const x = b64uToFixed32(jwk.x);
  const y = b64uToFixed32(jwk.y);
  return Buffer.concat([Buffer.from([0x04]), x, y]);
}

/**
 * ECDSA-SHA256 签名（RFC 6979 deterministic），输出 RAW 编码 r||s（64 字节）。
 *
 * @param privateKeyScalar P-256 私钥标量（32 字节 big-endian）
 * @param message 待签名消息原文
 * @returns 64 字节签名
 */
export function ecdsaSignRaw(
  privateKeyScalar: Uint8Array,
  message: Uint8Array,
): Uint8Array {
  if (privateKeyScalar.length !== 32) {
    throw new Error(`expected 32-byte P-256 private scalar, got ${privateKeyScalar.length}`);
  }
  // 显式 prehash=true（默认值），noble 会用 SHA-256 摘要。
  // lowS=false 与 Python/Go SDK 行为对齐（不强制低 S）。
  // format='compact' 是默认值，输出 64B = r||s。
  const sig = p256.sign(message, privateKeyScalar, {
    prehash: true,
    lowS: false,
    format: 'compact',
  });
  if (sig.length !== 64) {
    throw new Error(`unexpected signature length: ${sig.length}`);
  }
  return sig;
}

/**
 * ECDSA-SHA256 验签（RAW 64 字节签名）。
 */
export function ecdsaVerifyRaw(
  publicKeyDer: Uint8Array,
  signatureRaw: Uint8Array,
  message: Uint8Array,
): boolean {
  if (signatureRaw.length !== 64) return false;
  try {
    const uncompressed = derSpkiToUncompressed(publicKeyDer);
    return p256.verify(signatureRaw, message, uncompressed, {
      prehash: true,
      lowS: false,
      format: 'compact',
    });
  } catch {
    return false;
  }
}

/**
 * 仅用于本地校验：根据 P-256 私钥派生公钥的 DER SPKI 编码。
 */
export function privateScalarToPublicDer(privateKeyScalar: Uint8Array): Uint8Array {
  if (privateKeyScalar.length !== 32) {
    throw new Error(`expected 32-byte P-256 private scalar, got ${privateKeyScalar.length}`);
  }
  // noble 返回未压缩 65 字节
  const uncompressed = p256.getPublicKey(privateKeyScalar, false);
  if (uncompressed.length !== 65 || uncompressed[0] !== 0x04) {
    throw new Error('failed to derive uncompressed P-256 public key');
  }
  const x = uncompressed.subarray(1, 33);
  const y = uncompressed.subarray(33, 65);
  const pubKey = createPublicKey({
    key: {
      kty: 'EC',
      crv: 'P-256',
      x: Buffer.from(x).toString('base64url'),
      y: Buffer.from(y).toString('base64url'),
    },
    format: 'jwk',
  });
  return new Uint8Array(pubKey.export({ format: 'der', type: 'spki' }));
}
